From 1bbaadc4e2036a85b5a9f55c937ceed490d7e941 Mon Sep 17 00:00:00 2001
From: zachary <zhangzg@marvell.com>
Date: Wed, 28 Dec 2016 17:08:37 +0800
Subject: [PATCH 1235/2241] fix: spi: a3700: add spi data capture clock edge
 setting

- SPI data capture used to set positive edge for any frequency, which
  is not correct in case of SPI controller working in high frequency.
  If the data output delay from the flash is greater than 1/2 of the
  clock cycle (this is usually the case when running at high frequency)
  needs to set negative edge of the clock to capture data. For most of
  SPI flashes, the number is 7ns catpuring data time). So it means that
  SOC set clock edge to negative edge when SPI flash output flash
  frequency is more than 71MHz(1/14ns).
- This patch configures data capture edge according to SPI output
  frequencey.

Change-Id: Ic8425f67f93f3a96da2e6793669eb468914a4403
Signed-off-by: zachary <zhangzg@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/35092
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
---
 drivers/spi/spi-armada-3700.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/spi/spi-armada-3700.c b/drivers/spi/spi-armada-3700.c
index be24509..e8e2e23 100644
--- a/drivers/spi/spi-armada-3700.c
+++ b/drivers/spi/spi-armada-3700.c
@@ -95,6 +95,11 @@
 #define A3700_SPI_INSTR_CNT_BIT		0
 #define A3700_SPI_INSTR_CNT_MASK		0x3
 
+/* A3700_SPI_IF_TIME_REG */
+#define A3700_SPI_CLK_CAPT_EDGE		(1 << 7)
+
+#define DATA_CAP_CLOCK_CYCLE_NS		14	/*Minimum clock cycle of flash*/
+#define RVT_EDGE_FLASH_CAP_FREQ		(1000 * 1000 * 1000 / DATA_CAP_CLOCK_CYCLE_NS)
 
 struct a3700_spi_initdata {
 	unsigned int cs_num;
@@ -284,6 +289,19 @@ static int a3700_spi_clock_set(struct a3700_spi *a3700_spi,
 	val = val | (prescale & A3700_SPI_CLK_PRESCALE_MASK);
 	spireg_write(a3700_spi, A3700_SPI_IF_CFG_REG, val);
 
+	/*
+	* If the data output delay from the flash is greater than 1/2 of the clock cycle
+	* (this is usually the case when running at high frequency) needs to set negative
+	* edge of the clock to capture data. For most of SPI flashes, the number is 7ns
+	* (catpuring data time). So it means that SOC set CLK_CAPT_EDGE to negative
+	* edge when SPI flash output flash frequency is more than 71MHz(1/14ns).
+	*/
+	if (a3700_spi->input_clk_freq / prescale > RVT_EDGE_FLASH_CAP_FREQ) {
+		val = spireg_read(a3700_spi, A3700_SPI_IF_TIME_REG);
+		val |= A3700_SPI_CLK_CAPT_EDGE;
+		spireg_write(a3700_spi, A3700_SPI_IF_TIME_REG, val);
+	}
+
 	return 0;
 }
 
-- 
2.7.4

